### PR TITLE
CYS: allow to the user to move the pattern

### LINKED ISSUE #47325: [CYS - Full Composability]: Render a toolbar on top of the selected pattern
With Full Composability, users will be able to design their own homepage by adding, removing, shuffling, and rearranging patterns. These actions will be triggered thanks to a toolbar that will be visible on top of the selected pattern:

image

### PR DESCRIPTION
**Pull request by gigitux**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR allows the user to move blocks/patterns. 

https://github.com/woocommerce/woocommerce/assets/4463174/527c27ce-e9ba-454d-88f2-59ed8bd5420a

Closes #47325, #47326 and https://github.com/woocommerce/woocommerce/issues/47570

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure that you are using a Gutenberg trunk version
2. Make sure you have the WooCommerce beta tester plugin installed
3. Head over to Tools > WCA Test Helper > Features
4. Make sure you see the `pattern-toolkit-full-composability` on the list.
5. Enable it.
4. Head over to WooCommerce > Home > Customize your store.
5. Click on "Start designing" and proceed to the Pattern Assembler.
6. Click on "Design your Homepage".
7. Switch Pattern
8. Ensure that the block toolbar appears and you can move the block.
9. Ensure that you will not be able to move the block before the header and after the footer.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 

CYS: allow to the user to move the pattern.

#### Comment 

</details>

### CODE REVIEW

**Comment by albarin** [comment]
@gigitux Thanks for working on this, I'm seeing a few issues:
1. In the first template sidebar preview there are some errors.
2. When I change to the second template and come back to the first, some headers and texts have disappeared.
3. When entering the assembler the first time, there's no toolbar to move the patterns, it only appears when clicking on another template and coming back to the first.

https://github.com/woocommerce/woocommerce/assets/186112/c97d9a04-6bf3-4047-bd4c-e5cb8d3afd53

**Comment by gigitux** [comment]
> @gigitux Thanks for working on this, I'm seeing a few issues:
> 
> 1. In the first template sidebar preview there are some errors.
> 2. When I change to the second template and come back to the first, some headers and texts have disappeared.
> 3. When entering the assembler the first time, there's no toolbar to move the patterns, it only appears when clicking on another template and coming back to the first.
> 
>  CleanShot.2024-05-13.at.10.47.30.mp4

Thanks, @albarin, for the review!

1 & 2 are not specifically related to this PR. It looks like there is a regression on Gutenberg (you have the same issue if you try Gutenberg's trunk). We have to double-check!

> When entering the assembler the first time, there's no toolbar to move the patterns, it only appears when clicking on another template and coming back to the first.

The UX about the toolbar will be handled in another issue: https://github.com/woocommerce/woocommerce/issues/47371

**Review by albarin — APPROVED** [review]

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/block-preview.tsx [hunk 1]
```diff
@@ -17,6 +18,8 @@ import {
 	ScaledBlockPreviewProps,
 } from './auto-block-preview';
 import { ChangeHandler } from './hooks/use-editor-blocks';
+import { Toolbar } from './toolbar/toolbar';
+import { IS_TOOLBAR_ENABLED_FEATURE_FLAG } from './toolbar/feature-flag';
```
Comment (nefeline) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/block-preview.tsx]: Can you please update this feature flag to rely on our newly introduced `pattern-toolkit-full-composability` ? https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/admin/config/core.json#L22

**Review by nefeline — CHANGES_REQUESTED** [review]

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/block-preview.tsx (hunk 1 above)
Comment (gigitux) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/block-preview.tsx]: Yes! I updated it with [f58d81c](https://github.com/woocommerce/woocommerce/pull/47322/commits/f58d81ce32b99e421e12c4dfc3b44ccdb1acd3e7).

Let me know if I can merge this PR!

**Comment by nefeline** [comment]
> It looks like there is a regression on Gutenberg (you have the same issue if you try Gutenberg's trunk). We have to double-check!

Could we investigate if this issue is indeed a consequence of a regression on GB and not from this PR?

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/block-preview.tsx (hunk 1 above)
Comment (nefeline) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/block-preview.tsx]: Thanks! I've left some additional questions related to the issues found during the code review.

**Comment by gigitux** [comment]
> Could we investigate if this issue is indeed a consequence of a regression on GB and not from this PR?

I checked the WooCommerce trunk branch, and I'm able to reproduce the issue :)

**Comment by nefeline** [comment]
Thanks for the updates @gigitux ! Sharing here my test results:

- Whenever accessing the editor with the GB plugin disabled, I'm greeted with an error:

Screenshot 2024-05-16 at 13 09 09

Could we make a change to prevent this?

> I checked the WooCommerce trunk branch, and I'm able to reproduce the issue :)

Cool, thanks: do you think its worth creating issues for these two? Maybe you already did, but flagging anyways just to make sure we are on the same page:

> - In the first template sidebar preview there are some errors.
> - When I change to the second template and come back to the first, some headers and texts have disappeared.

**Comment by gigitux** [comment]
> In the first template sidebar preview there are some errors.

This one is tracked by https://github.com/woocommerce/woocommerce/issues/47469!

> When I change to the second template and come back to the first, some headers and texts have disappeared.

I'm not able to replicate it:
https://github.com/woocommerce/woocommerce/assets/4463174/2b247dc3-887e-47c2-a3c5-3b4f083f161b

> Whenever accessing the editor with the GB plugin disabled, I'm greeted with an error:

I created an issue: https://github.com/woocommerce/woocommerce/issues/47545. I'm going to prioritize this!

**Comment by nefeline** [comment]
Thanks!

> I'm not able to replicate it:

Hm, yeah, I can't replicate it either. But I can replicate on this PR the same issue observed on https://github.com/woocommerce/woocommerce/pull/47356 regarding the scroll:

https://github.com/woocommerce/woocommerce/assets/15730971/4062ac47-95d9-4ee5-89f3-6c3e2d824b32

Just to confirm: do we have an issue open to solve this problem, in particular? If we do and we can prioritize it, I'm a-ok with approving/merging this PR 👌

**Comment by nefeline** [comment]
Ah: can you also please update the description of this PR so it accounts for the changes on the feature flag & the fact that the PR on GB has already been merged? Thanks!

**Comment by gigitux** [comment]
> Just to confirm: do we have an issue open to solve this problem, in particular? If we do and we can prioritize it, I'm a-ok with approving/merging this PR 👌

I created a dedicated issue: https://github.com/woocommerce/woocommerce/issues/47570 - are you able to reproduce this issue on https://github.com/woocommerce/woocommerce/pull/47415?


> Ah: can you also please update the description of this PR so it accounts for the changes on the feature flag & the fact that the PR on GB has already been merged? Thanks!

Done! Thanks for the review!

**Comment by nefeline** [comment]
> Done! Thanks for the review!

Thanks @gigitux ! Do you mind double-checking the testing instructions to see if they are accurate? I'm still seeing references to IS_TOOLBAR_ENABLED_FEATURE_FLAG.

**Comment by nefeline** [comment]
Thanks for updating the PR testing instructions again 🙏 

The notes I have here are very similar to the ones I shared on https://github.com/woocommerce/woocommerce/pull/47356#pullrequestreview-2068526586

1- I think this was mentioned here earlier, but whenever clicking on "Design your homepage" for the very first time, the toolbar to move patterns is not displayed on the first template.
2. Even after the user leaves the "Design your homepage" section, they can still see the tooltip to move the patterns.

I'm ok with addressing those separately, as long as we have issues to solve the aforementioned problems.

Regarding the scroll problem:

> are you able to reproduce this issue on https://github.com/woocommerce/woocommerce/pull/47415?

No, I'm not able to reproduce on #47415 : does this mean that you solved the problem via 47415? If yes, it would be nice to document those things here on this PR so the reviewers are aware of known issues and how they are being solved.

**Comment by gigitux** [comment]
> No, I'm not able to reproduce on #47415 : does this mean that you solved the problem via 47415? If yes, it would be nice to document those things here on this PR so the reviewers are aware of known issues and how they are being solved.

I changed nothing in the specific, so I'm not sure. It is better to keep the issue open.

> 1- I think this was mentioned here earlier, but whenever clicking on "Design your homepage" for the very first time, the toolbar to move patterns is not displayed on the first template.

https://github.com/woocommerce/woocommerce/issues/47752

> 2. Even after the user leaves the "Design your homepage" section, they can still see the tooltip to move the patterns.

https://github.com/woocommerce/woocommerce/issues/47751


Created!

**Comment by nefeline** [comment]
> I changed nothing in the specific, so I'm not sure. It is better to keep the issue open.

Do you think you can determine whether this PR introduces this bug? I'm concerned about approving it, merging it as-is, and then not being able to scroll up and down in the frame anymore. This type of bug can block us from doing other unrelated work down the line.

**Comment by gigitux** [comment]
> > I changed nothing in the specific, so I'm not sure. It is better to keep the issue open.
> 
> Do you think you can determine whether this PR introduces this bug? I'm concerned about approving it, merging it as-is, and then not being able to scroll up and down in the frame anymore. This type of bug can block us from doing other unrelated work down the line.

I fixed it with 15564d2b206f4625c0daceca7450a2b5efcce639

**Comment by nefeline** [comment]
> I fixed it with https://github.com/woocommerce/woocommerce/commit/15564d2b206f4625c0daceca7450a2b5efcce639

Thanks: I did a test, but unfortunately, this didn't fix the bug: it is still impossible to scroll up and down. To reproduce click on the second template: as soon as the taskbar with the move button is displayed, the frame freezes, and it is impossible to scroll.

**Review by nefeline — APPROVED** [review]
Registering here for future reference that additional tests were made, and of 3 different WordPress installs, I was able to repro the issue in one of them. https://github.com/woocommerce/woocommerce/issues/47855 has been created to keep an eye on this as there might be certain conditions to trigger it that were just not uncovered yet, given the consistency of the freeze.

**Pull request merged**